pkg/liqo-controller-manager/namespaceMap-controller/namespacemap_controller.go

```diff
@@ -29,61 +30,69 @@ import (
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		return ctrl.Result{}, err
 	}
 
-	if !ctrlutils.ContainsFinalizer(namespaceMap, liqoconst.NamespaceMapControllerFinalizer) {
-		ctrlutils.AddFinalizer(namespaceMap, liqoconst.NamespaceMapControllerFinalizer)
-		if err := r.Patch(context.TODO(), namespaceMap, client.Merge); err != nil {
-			klog.Errorf("%s --> Unable to add finalizer to the NamespaceMap '%s'", err, namespaceMap.GetName())
+	// If the NamespaceMap is requested to be deleted
+	if !namespaceMap.DeletionTimestamp.IsZero() {
+		if err := r.namespaceMapDeletionProcess(ctx, namespaceMap); err != nil {
 			return ctrl.Result{}, err
 		}
-		klog.Infof("Finalizer correctly added on NamespaceMap '%s'", namespaceMap.GetName())
+		return ctrl.Result{}, nil
 	}
 
-	if err := r.manageRemoteNamespaces(namespaceMap); err != nil {
+	// If someone deletes namespaceMap is necessary to remove all remote namespace
+	// associated with this resource before deleting it
+	if err := r.SetNamespaceMapControllerFinalizer(ctx, namespaceMap); err != nil {
 		return ctrl.Result{}, err
 	}
 
-	if len(namespaceMap.Status.CurrentMapping) == 0 {
-		ctrlutils.RemoveFinalizer(namespaceMap, liqoconst.NamespaceMapControllerFinalizer)
-		if err := r.Update(context.TODO(), namespaceMap); err != nil {
-			klog.Errorf("%s --> Unable to remove '%s' from NamespaceMap '%s'", err,
-				liqoconst.NamespaceMapControllerFinalizer, namespaceMap.GetName())
-			return ctrl.Result{}, err
-		}
-		klog.Infof("Finalizer correctly removed from NamespaceMap '%s'", namespaceMap.GetName())
+	// Create/Delete remote Namespaces if it is necessary, according to NamespaceMap status.
+	if err := r.ensureRemoteNamespaces(ctx, namespaceMap); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Only when all remote namespaces associated with this NamespaceMap are deleted, remove finalizer.
+	if _, err := r.TryToRemoveNamespaceMapControllerFinalizer(ctx, namespaceMap); err != nil {
+		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
-// The Controller is triggered only when the number of entries in DesiredMapping changes, so only when
-// a namespace's request is added or removed.
+// Events not filtered:
+// 1 -- the number of entries in DesiredMapping changes, so a namespace's request is added or removed.
+// 2 -- the NamespaceMap is deleted and has the NamespaceMapControllerFinalizer.
 func manageDesiredMappings() predicate.Predicate {
 	return predicate.Funcs{
 		UpdateFunc: func(e event.UpdateEvent) bool {
+			// If the NamespaceMap is deleted and has my finalizer.
+			if !e.ObjectNew.GetDeletionTimestamp().IsZero() &&
+				ctrlutils.ContainsFinalizer(e.ObjectNew, namespaceMapControllerFinalizer) {
+				return true
+			}
+			// If an entry is added or removed in DesiredMapping
 			return len(e.ObjectOld.(*mapsv1alpha1.NamespaceMap).Spec.DesiredMapping) !=
 				len(e.ObjectNew.(*mapsv1alpha1.NamespaceMap).Spec.DesiredMapping)
```

Comment:
Shouldn't you also perform a reconcile if a mapping is modified (and the overall array length remains the same)?

Reply:
yes, you are right maybe an entry is removed and another is added immediately when the event is evaluated the length is the same